### PR TITLE
add ethereum expansion logic

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**

### CODE REVIEW

File: eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthDepositInitialization.kt [hunk 1]
```diff
@@ -44,19 +53,58 @@ class EthDepositInitialization(
     private val irohaAPI: IrohaAPI,
     private val ethDepositConfig: EthDepositConfig,
     private val passwordsConfig: EthereumPasswords,
+    rmqConfig: RMQConfig,
     private val ethRelayProvider: EthRelayProvider,
     private val ethTokensProvider: EthTokensProvider
 ) {
+    private var ecKeyPair: ECKeyPair =
+        DeployHelper(ethDepositConfig.ethereum, passwordsConfig).credentials.ecKeyPair
```
Comment (Mingela) [inline eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthDepositInitialization.kt]: how about adding a static method to `DeployHelper` to do it?
```
WalletUtils.loadCredentials(
            ethereumPasswords.credentialsPassword,
            ethereumConfig.credentialsPath
        )
```

File: eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthDepositInitialization.kt [hunk 2]
```diff
@@ -44,19 +53,58 @@ class EthDepositInitialization(
...
         return initEthChain()
-            .map { ethEvent ->
-                initNotary(ethEvent)
-            }
+            .map { ethEvent -> initNotary(ethEvent) }
             .flatMap { notary -> notary.initIrohaConsumer() }
+            .flatMap { irohaChainListener.getBlockObservable() }
+            .flatMap { irohaObservable ->
+                irohaObservable
+                    .subscribeOn(
```
Comment (Mingela) [inline eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthDepositInitialization.kt]: `observeOn`

File: eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthereumDepositExpansionStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                    String.unHex(expansionDetails.publicKey.toLowerCase())
+                )
+                .setAccountQuorum(notaryCredential.accountId, expansionDetails.quorum)
+                .setQuorum(consumer.getConsumerQuorum().get())
+                .setAccountDetail(
+                    ethDepositConfig.notaryListStorageAccount,
+                    notaryName,
+                    notaryEndpoint
+                )
+                .setCreatedTime(triggerTime)
```
Comment (Mingela) [inline eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthereumDepositExpansionStrategy.kt]: it is possible to set `created time` right in `builder` constructor

File: eth-deposit/src/main/kotlin/com/d3/eth/deposit/endpoint/EthAddPeerStrategy.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * Check and give proof
+     */
+    override fun performAddPeer(irohaTxHash: IrohaTransactionHashType): EthNotaryResponse {
+        val irohaTx = queryHelper.getSingleTransaction(irohaTxHash)
+
+        return irohaTx.map { tx ->
+            if (tx.payload.reducedPayload.creatorAccountId != expansionTriggerCreatorAccountId)
+                throw IllegalArgumentException("Wrong iroha tx creator account id ${tx.payload.reducedPayload.creatorAccountId}, expected ${expansionTriggerCreatorAccountId}")
+            val setAccountDetail = tx.payload.reducedPayload.getCommands(0).setAccountDetail
```
Comment (Mingela) [inline eth-deposit/src/main/kotlin/com/d3/eth/deposit/endpoint/EthAddPeerStrategy.kt]: is it mandatory for this transaction to have setAccountDetails command and must it be the first one?

File: eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt [hunk 5]
```diff
@@ -43,6 +47,52 @@ class WithdrawalServiceInitialization(
...
+        withdrawalConfig.irohaAnchoredTokenSetterAccount
+    )
+
+    private val notaryPeerListProvider = NotaryPeerListProviderImpl(
+        queryHelper,
+        withdrawalConfig.notaryListStorageAccount,
+        withdrawalConfig.notaryListSetterAccount
+    )
+
+
```
Comment (Mingela) [inline eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt]: one line should be enough 🐤

File: eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt [hunk 6]
```diff
@@ -43,6 +47,52 @@ class WithdrawalServiceInitialization(
...
+    val ethConsumer = EthConsumer(
+        withdrawalConfig.ethereum,
+        withdrawalEthereumPasswords,
+        relayVacuumConfig
+    )
+
+
```
Comment (Mingela) [inline eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt]: one line should be enough 🐤

File: eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt [hunk 7]
```diff
@@ -51,26 +101,32 @@ class WithdrawalServiceInitialization(
         logger.info { "Init Iroha chain listener" }
         return chainListener.getBlockObservable()
             .map { observable ->
-                observable.flatMapIterable { (block, _) -> IrohaChainHandler().parseBlock(block) }
+                observable.flatMapIterable { (block, _) ->
+                    ethereumWithdrawalExpansionStrategy.filterAndExpand(block)
+                    IrohaChainHandler().parseBlock(block)
```
Comment (Mingela) [inline eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt:104]: it is already done in `filterAndExpand(block)`, isn't it?

Comment (Alexey-N-Chernyshov) [reply eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt:104]: No, ethereumWithdrawalExpansionStrategy.filterAndExpand(block) responsible only for expansion, the rest is performed by parseBlock()

Comment (Mingela) [reply eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt:104]: Why do you have `IrohaChainHandler().parseBlock(block)` inside `filterAndExpand()`
EthereumWithdrawalExpansionStrategy:51

Comment (Alexey-N-Chernyshov) [reply eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceInitialization.kt:104]: fixed

**Review by Mingela — APPROVED** [review]

File: eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthereumDepositExpansionStrategy.kt [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    )
+
+    /**
+     * Filter expansion trigger event and call expansion logic
+     * @param block - iroha block
+     */
+    fun filterAndExpand(block: BlockOuterClass.Block) {
+        expansionService.expand(block) { expansionDetails, _, triggerTime ->
+            if (expansionDetails.accountIdToExpand == notaryCredential.accountId) {
+                EthDepositInitialization.logger.info { "Add new peer with iroha pubkey ${expansionDetails.publicKey} endpoint ${expansionDetails.additionalData["notary_endpoint"]}" }
```
Comment (dolgopolovwork) [inline eth-deposit/src/main/kotlin/com/d3/eth/deposit/EthereumDepositExpansionStrategy.kt]: Logger from another file

**Pull request merged**